lib/policy/policyValidator.js

```diff
@@ -0,0 +1,117 @@
+'use strict'; // eslint-disable-line strict
+
+const Ajv = require('ajv');
+const userPolicySchema = require('./userPolicySchema');
+const errors = require('../errors');
+
+const ajValidate = new Ajv({ allErrors: true });
+// compiles schema to functions and caches them for all cases
+const userPolicyValidate = ajValidate.compile(userPolicySchema);
+
+const errDict = {
+    required: {
+        Version: 'Policy document must be version 2012-10-17 or greater.',
+        Action: 'Policy statement must contain actions.',
+    },
+    pattern: {
+        Action: 'Actions/Conditions must be prefaced by a vendor,' +
+            ' e.g., iam, sdb, ec2, etc.',
+        Resource: 'Resource must be in ARN format or "*".',
+    },
+    minItems: {
+        Resource: 'Policy statement must contain resources.',
+    },
+};
+
+// parse ajv errors and build list of erros
+function _parseErrors(ajvErrors) {
+    // deep copy is needed as we have to assign custom error description
+    const parsedErr = Object.assign({}, errors.MalformedPolicyDocument);
+    parsedErr.description = 'Syntax errors in policy.';
+    ajvErrors.some(err => {
+        const resource = err.dataPath;
+        const field = err.params ? err.params.missingProperty : undefined;
+        const errType = err.keyword;
+        if (errType === 'type' && resource === '.Statement') {
+            // skip this as this doesn't have enough error context
+            return false;
+        }
+        if (err.keyword === 'required' && field && errDict.required[field]) {
+            parsedErr.description = errDict.required[field];
+            return true;
+        }
+        if (err.keyword === 'pattern' && (resource === '.Statement.Action' ||
+            resource === '.Statement.NotAction')) {
+            parsedErr.description = errDict.pattern.Action;
+            return true;
+        }
+        if (err.keyword === 'pattern' && (resource === '.Statement.Resource' ||
+            resource === '.Statement.NotResource')) {
+            parsedErr.description = errDict.pattern.Resource;
+            return true;
+        }
+        // exit with general error
+        return true;
+    });
+    return parsedErr;
+}
+
+
+// parse JSON safely without throwing an exception
+function _safeJSONParse(s) {
+    try {
+        return JSON.parse(s);
+    } catch (e) {
+        return e;
+    }
+}
+
+
+// validates policy using the validation schema
+function _validatePolicy(type, policy) {
+    if (type === 'user') {
+        const parseRes = _safeJSONParse(policy);
+        if (parseRes instanceof Error) {
+            return { error: Object.assign({}, errors.MalformedPolicyDocument),
+                valid: false };
+        }
+        userPolicyValidate(parseRes);
+        if (userPolicyValidate.errors) {
```

Comment:
Their API feels strange, having one global function that monkey-patches itself. Seems OK here but feels so unsafe 😭